### PR TITLE
new_ghost shouldn't clear newargs (PyPy consistent with C)

### LINKED ISSUE #49: PyPy/Pure-Python: `PickleCache.new_ghost` destroys any information passed to `__new__`
Spurred from the work in #44, with a lengthy description in https://github.com/zopefoundation/persistent/pull/44#issuecomment-256962199

Both dicts and slots are affected and need a test case.

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #49. Tested with ZODB, ZEO, zope.container.

(Includes #47 , rebased on master and tested with PyPy on ZEO and ZODB, because when I created this branch I expected that PR to be merged already by the time this was ready.)

### CODE REVIEW

File: persistent/tests/test_persistence.py [hunk 1]
```diff
@@ -1627,6 +1675,10 @@ def update_object_size_estimation(self, oid, new_size):
 
         return _Cache(jar)
 
+    def _makeRealCache(self, jar):
+        from persistent.picklecache import PickleCache
```
Comment (jamadden) [inline persistent/tests/test_persistence.py]: CPersistentTests already used cPickleCache in _makeCache, not a mock like the PyPersistentTests did.

File: persistent/persistence.py [hunk 2]
```diff
@@ -413,27 +413,29 @@ def _p_invalidate(self):
...
+            # for backward-compatibility reason we release __slots__ only if
+            # class does not override __new__
+            if type_.__new__ is Persistent.__new__:
+                for slotname in Persistent._slotnames(self, _v_exclude=False):
+                    try:
+                        getattr(type_, slotname).__delete__(self)
+                    except AttributeError:
+                        # AttributeError means slot variable was not initialized at all -
+                        # - we can simply skip its deletion.
+                        pass
```
Comment (tseaver) [inline persistent/persistence.py]: ISTM that we need a protocol by which a class derived from `Persistent` can signal signal explicitly which attrs / slots should be preserved:  all the others *should* be cleared by `new_ghost` (a ghost isn't supposed to have any persistent state set).

Comment (jamadden) [inline persistent/persistence.py]: This was sort-of talked over in #44. I think the consensus was that some sort of protocol would be nice, BWC makes it difficult to do.

The change here just brings parity between the Python implementation and what the C implementation is already doing.

**Comment by jamadden** [comment]
@tseaver Any further comments?

**Comment by tseaver** [comment]
I'm still puzzled that the C version of `new_ghost` doesn't clear object state:  the whole `__getattr__`-triggers-load scheme kind of depends on it.

**Comment by jamadden** [comment]
The only state an object would have when `new_ghost` is called by ZODB is what comes from `__getnewargs__`. Most of the C implemented objects store that information in private `struct` fields, not in slots or the dict (what Jim called "intrinsic data"), so it wouldn't have been cleared *anyway*. 

The situation for Python-implemented objects is more complex, but with the C cache implementation, nothing passed to `__new__` would have been cleared at `new_ghost` time. The Python cache implementation was behaving differently, breaking objects that were relying on that. 

I don't know any *actual* cases where this implementation difference has mattered, but the increasing complexity of how `__slots__` are handled---not ever being cleared if `__new__` is overridden---suggests that we at least need the test cases for this behaviour to make sure that it is consistent between the implementations and doesn't *accidentally* regress.

**Comment by tseaver** [comment]
LGTM

**Comment by jamadden** [comment]
Thanks for the review!

**Pull request merged**